### DISCUSSION TITLE
FSM 사용과 상태변경

### DISCUSSION THREAD

**Discussion by leehosung** [8percent/styleguide#25] [Ideas]
제안사항 

- Django Model 의 모든 상태 (state, status, progress 등) 에 FSM 을 의무적으로 적용합니다. 특별한 이유가 있지 않는 한.
- 모든 상태변경을 Model 내의 함수를 통해서 하도록 하고, FSM decorator를 통해 이를 통제합니다. 
![image](https://user-images.githubusercontent.com/2469766/154884614-93876748-a119-448b-b00d-3da5bd95ec66.png)
- 현재 사용하고 있는 패키지: https://github.com/viewflow/django-fsm 
- Reference
  - https://medium.com/swlh/finite-state-machines-with-django-b44198e1ca13

**Comment by Kirade** [comment]
동의합니다.

비즈니스 로직의 의도를 명시적으로 지정해주고, 강제적으로 규칙을 정해줄 수 있어
데이터 정합성 문제와 운영 문제를 줄이는데 도움이 될 것 같습니다.

**Comment by leehosung** [comment]
@Kirade 혹시 djang-fsm 이 취하는 디자인 패턴이 저희와 일치 하지 않는다고 말씀 주신 부분은 어떤걸까요?

**Reply by Kirade** [reply]
https://github.com/viewflow/django-fsm/issues/79

이 부분 인데요. 
이전에 간단히 논의했듯, 정의한 transition method 는 save()를 동시해 수행해주지 않기 때문에 save 를 별도로 호출해주어야 합니다.
저희가 https://github.com/8percent/styleguide/discussions/12 에서 논의한 내용과 반대되는 내용입니다.

이슈 내용을 조금 내려보니 트렌지션 내부에서 save 호출하는것으로 대안을 제시하고 있는것으로 보이는데요.
이것이 가능하면 문제없을것 같습니다.

**Reply by leehosung** [reply]
네. 그렇겠네요.

**Comment by leehosung** [comment]
https://github.com/jazzband/django-fsm-log

여기를 보면 state 변경에 따른 logging 도 잘 사용할수 있을것 같네요.

**Reply by Kirade** [reply]
tracking 모델을 대체할 수 있을지도 모르겠네요.

**Reply by leehosung** [reply]
네 맞아요.
